### ISSUE TITLE
BUG: ``np.dtype`` object does not produce consistent hashes.

### ISSUE THREAD

**Issue by theXYZT** [numpy/numpy#17864]
It seems that `np.dtype(...)` does not produce consistent hashes and does not equal the hash of the contained dtype. I feel like these should be consistent with each other so that the following will return true:

```
np.empty(1).dtype in {np.float32, np.float64}  # False
```

### Reproducing code example:

```
In [1]: np.float64 == np.dtype(np.float64)
Out[1]: True

In [2]: np.float64 in {np.dtype(np.float64)}
Out[2]: False

In [3]: hash(np.float64)
Out[3]: 8793996338852

In [4]: hash(np.dtype(np.float64))
Out[4]: 3855163183283070272
```

`hash(np.dtype(np.float64))` returns a different value when called in different runtimes too.

---

Would this be worth fixing?

**Comment by asmeurer** [comment]
This behavior also leads to surprising behavior if you try to use dtypes as dictionary keys. Generally in Python, two objects that are equal are supposed to have the same hash. Any reason that `dtype.__hash__` can't be set to `hash(dtype.type)` whenever the dtype equals `dtype.type`?

**Comment by rkern** [comment]
We've certainly [noted and discussed this before](https://groups.google.com/g/numpy/c/e3Vvj6MkOL0). The main problem with following the general rule is that `dtype.__eq__` is _very_ promiscuous. `dtype.__hash__` _can't_ follow the general rule that "objects that compare equal must hash equal" because we've designed `dtype.__eq__` to make many different objects that already have incompatible `__hash__` implementations compare equal. I can't make `hash(dtype(np.float64)) == hash('float64') and hash(dtype('f8')) == hash('f8')` even though `dtype(np.float64) == 'float64' and dtype(np.float64) == 'f8'`.

I'm not sure about the proposed rule. It's possible that it's reasonable. But given how wonky the comparison and hashing semantics are for `np.dtype`, I'm not sure that I want it to work for just a few more common cases. The danger is that users try out the obvious things, see that they work, and carry on with a mental model that generalizes that to a lot of cases where it doesn't work. My general recommendation right now is to just treat `dtype` instances as unhashable.

**Comment by asmeurer** [comment]
I guess for me it's a lot less surprising that `{np.float64: ...}['f8']` raises KeyError than that `{np.float64: ...}[np.array(..., dtype=np.float64).dtype]` does. In particular, I don't think that many people may even realize that there are two different types of "dtype objects" for each dtype, the dtype class and the dtype instance.

**Comment by rkern** [comment]
Well, `np.float64` and friends _aren't_ "dtype classes". There's no such thing. They are scalar types, instances of which are scalar instances which have certain dtypes that are associated with the scalar type. We've ended up using them as specifying the desired dtype (e.g. `dtype=np.float32` arguments), but that's just a convention that we added on because it was a convenient alternative to passing around strings and we had a cluttered namespace already. We made them do double-duty.

Basically, we created a Frankenstein's monster of complicated semantics. I don't think that adding a patch-up here actually simplifies anything. It just makes it harder for people to run into the problems with superficial tests, which is how a lot of people build the mental models of how things work; systematically trying counterexamples is hard work, and we don't do it often. But it doesn't necessarily make it harder to run into the problems later. I lean towards surprising early rather than late.

For these kinds of uses (distinguishing between the basic, native-endian numeric dtypes), I recommend keying off of `dtype.name`.

**Comment by asmeurer** [comment]
My use-case is the array API, where `xp.float64` and `array.dtype` are the only way to access dtypes. So like it or not, that is the best way to specify dtypes if you want to write cross-library compatible code. 

The array API doesn't require any methods on dtypes except for `__eq__`, which does strictly mean we can't rely on hashing either, but NumPy is the only library that actually has issues with hashing dtype objects.

**Comment by rkern** [comment]
Ah, that's a good point. That could be a useful well-defined boundary to provide scope for a practical mental model.

We could, in principle, make `hash(some_dtype) == hash(some_dtype.type)` always (even in the cases where `some_dtype != np.dtype(some_dtype.type)` like non-native endian or variable-length dtypes). It's a poor (but valid!) hash function per se (in principle, it could allow certain attacks by causing quadratic behavior in dictionaries; not sure anyone considers that CVE-worthy, but all the same). There _might_ be some complications to enforcing your proposed semantics because both the builtin dtype objects and the scalar types are defined statically; there might be ordering issues, but in principle, it can be done.

If we did this, I'd like to see a concurrent addition to the array API that enforces the hash semantics for the [named API-standard dtype specifiers](https://data-apis.org/array-api/latest/API_specification/data_types.html). I suspect that it was omitted because numpy was the oddball implementation and no one wanted to touch it. But if we did this in concert, I'd support it.

**Comment by jorenham** [comment]
> I've seen often recited that objects that compare equal and are hashable must have the same hash [citation needed].

It's the other way around: If two objects hash to the same value, then they must be equal. 

~There is no general rule that says that the converse must also hold.~ However, I agree that this is pretty intuitive behavior, and can easily lead to bugs, e.g. by applying ruff's [`PLR1714`](https://docs.astral.sh/ruff/rules/repeated-equality-comparison/) rule.

In this case, we're comparing a `dtype` **object** to a `float64` **type**. They're fundamentally incompatible, so should not hash to the same value. And although I can see the convenience in them being comparable through `__eq__`, it blurs the lines between dtypes objects and scalar types, making it even more confusing than it already is. 
So personally, I would rather have them not be comparable. However, changing this behavior would probably not be easy, to say the least.

**Comment by ricardoV94** [comment]
> It's the other way around: If two objects hash to the same value, then they must be equal.

That's not correct. Two objects can have the same hash and not be equal, that's just a hash collision.

**Comment by ricardoV94** [comment]
> However, even Python itself is fundamentally built upon the idea that this will not happen

Why do you say this? Dictionary / set operations use equality to disambiguate objects with the same hash, just try to store two things in a dict and you'll see the two methods being called. 

Hash is a cheap smoke test for equality, and it has to be consistent with it, but it doesn't need to be unique over unequal items. It's just better if it is, for performance reasons not correctness. I never heard that python assumes otherwise...

**Comment by jorenham** [comment]
> Why do you say this?

Because otherwise we would not have `set` or `dict`.

**Comment by ricardoV94** [comment]
You're wrong: https://docs.python.org/3/glossary.html#term-hashable

That's my missed citation from my first comment as well

**Comment by ricardoV94** [comment]
Either equality behavior should change or numpy dtypes shouldn't be hashable, perhaps the latter is less painful for users?

Some discussion in https://stackoverflow.com/questions/35293672/why-do-these-dtypes-compare-equal-but-hash-different

**Comment by jorenham** [comment]
> You're wrong: [docs.python.org/3/glossary.html#term-hashable](https://docs.python.org/3/glossary.html#term-hashable)
> 
> That's my missed citation from my first comment as well
> 
> > Hashable objects which compare equal must have the same hash value.

Yea it appears that I was wrong about there not being such a rule; I stand corrected. This is then a bit more serious that I initially thought it would be. Thanks for noticing.

Like I said in https://github.com/numpy/numpy/issues/17864#issuecomment-2925796851, I think it would be best to keep their hashes different, and change `__eq__` so that `dtype(np.float64) != np.float64`. In fact, I think it's the only way that's realistically possible: 
Deprecating `dtype.__eq__` for scalar types is possible, but deprecating comparing the *hashes* of dtype objects and scalar types is not. If we would make their hash functions the same, it would break a lot of code. For instance `{np.dtype(float), np.float64}` would then result in either `{np.dtype(float)}` or `{np.float64}`.
